# Basic data analysis in R

R is a statistical programming language that is purpose built for data analysis. 

Base R does a lot, but there are a mountain of external libraries that do things to make R better/easier/more fully featured. One of the best libraries, in your professor's opinion, is `dplyr`, a library for working with data. To use dplyr, you need to import it. 

In [1]:
library(dplyr)


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



The first thing we need to do is get some data to work with. We do that by reading it in. In our case, we're going to read data from a csv file -- a comma-separated values file. 

The code looks like this: 

`mountainlions <- read.csv("../../Data/mountainlions.csv")`

Let's unpack that. 

The first part -- `mountainlions` -- is the name of your variable. A variable is just a name of a thing. In this case, our variable is a data frame, which is R's way of storing data. We can call this whatever we want. I always want to name data frames after what is in it. In this case, we're going to import a dataset of mountain lion sightings from the Nebraska Game and Parks Commission.

The `<-` bit is the variable assignment operator. It's how we know we're assigning something to a word. 

The `read.csv` bits are pretty obvious. What happens in the quote marks is the path to the data. In there, I have to tell R where it find the data. The easiest thing to do, if you are confused about how to find your data, is to put your data in the same folder as as your notebook. In my case, I've got a folder called Data that's two levels up from my work folder. So the `../` means move up one level. So move up one level, move up one level, find Data, then in there is a file called mountainlions.csv. 

What you put in there will be different from mine. So your first task is to import the data. 

In [2]:
mountainlions <- read.csv("../../Data/mountainlions.csv")

Now we can inspect the data we imported. What does it look like? To do that, we use `head(mountainlions)` to show the headers and the first six rows of data. If we wanted to see them all, we could just simply enter `mountainlions` and run it. 

To get the number of records in our dataset, we run `nrow(mountainlions)`

In [3]:
head(mountainlions)
nrow(mountainlions)

ID Cofirm.Type COUNTY       Date    
1 1  Track       Dawes        9/14/91 
2 2  Mortality   Sioux        11/10/91
3 3  Mortality   Scotts Bluff 4/21/96 
4 4  Mortality   Sioux        5/9/99  
5 5  Mortality   Box Butte    9/29/99 
6 6  Track       Scotts Bluff 11/12/99

[1] 393

So what if we wanted to know how many mountain lion sightings there were in each county? To do that by hand, we'd have to take each of the 393 records and sort them into a pile. We'd put them in groups and then count them. 

`dplyr` has a group by function in it that does just this. A massive amount of data analysis involves grouping like things together at some point. So it's a good place to start. 

So to do this, we'll take our dataset and we'll introduce a new operator: `%>%`. The best way to read that operator, in my opinion, is to interpret that as "and then do this." Here's the code: 

In [4]:
mountainlions %>%
  group_by(COUNTY) %>%
  summarise(
    count = n(),
  ) 

COUNTY       count
1  Banner         6  
2  Blaine         3  
3  Box Butte      4  
4  Brown         15  
5  Buffalo        3  
6  Cedar          1  
7  Cherry        30  
8  Custer         8  
9  Dakota         3  
10 Dawes        111  
11 Dawson         5  
12 Dixon          3  
13 Douglas        2  
14 Frontier       1  
15 Hall           1  
16 Holt           2  
17 Hooker         1  
18 Howard         3  
19 Keith          1  
20 Keya Paha     20  
21 Kimball        1  
22 Knox           8  
23 Lincoln       10  
24 Merrick        1  
25 Morrill        2  
26 Nance          1  
27 Nemaha         5  
28 Platte         1  
29 Polk           1  
30 Richardson     2  
31 Rock          11  
32 Sarpy          1  
33 Saunders       2  
34 Scotts Bluff  26  
35 sheridan       2  
36 Sheridan      35  
37 Sherman        1  
38 Sioux         52  
39 Thomas         5  
40 Thurston       1  
41 Valley         1  
42 Wheeler        1

So let's walk through that. We start with our dataset -- `mountainlions` -- and then we tell it to group the data by a given field in the data. In this case, we wanted to group together all the counties, signified by the field name COUNTY, which you could get from looking at `head(mountainlions)`. So after we group the data, we need to count them up. In dplyr, we use `summarize` [which can do more than just count things](http://dplyr.tidyverse.org/reference/summarise.html). So inside the parentheses in summarize, we set up the summaries we want. In this case, we just want a count of the counties. So `count = n(),` says create a new field, called `count` and set it equal to `n()`, which might look weird, but it's common in stats. The number of things in a dataset? Statisticians call in n. There are n number of incidents in this dataset. So `n()` is a function that counts the number of things there are. 

And when we run that, we get a list of counties with a count next to them. But it's not in any order. So we'll add another And Then Do This %>% and use `arrange`. Arrange does what you think it does -- it arranges data in order. By default, it's in ascending order -- smallest to largest. But if we want to know the county with the most mountain lion sightings, we need to sort it in descending order. That looks like this:

In [5]:
mountainlions %>%
  group_by(COUNTY) %>%
  summarise(
    count = n(),
  ) %>% arrange(desc(count))

COUNTY       count
1  Dawes        111  
2  Sioux         52  
3  Sheridan      35  
4  Cherry        30  
5  Scotts Bluff  26  
6  Keya Paha     20  
7  Brown         15  
8  Rock          11  
9  Lincoln       10  
10 Custer         8  
11 Knox           8  
12 Banner         6  
13 Dawson         5  
14 Nemaha         5  
15 Thomas         5  
16 Box Butte      4  
17 Blaine         3  
18 Buffalo        3  
19 Dakota         3  
20 Dixon          3  
21 Howard         3  
22 Douglas        2  
23 Holt           2  
24 Morrill        2  
25 Richardson     2  
26 Saunders       2  
27 sheridan       2  
28 Cedar          1  
29 Frontier       1  
30 Hall           1  
31 Hooker         1  
32 Keith          1  
33 Kimball        1  
34 Merrick        1  
35 Nance          1  
36 Platte         1  
37 Polk           1  
38 Sarpy          1  
39 Sherman        1  
40 Thurston       1  
41 Valley         1  
42 Wheeler        1

## More basics

In the last example, we grouped some data together and counted it up, but there's so much more you can do. You can do multiple measures in a single step as well. 

Let's look at some simple college data. 

In [10]:
colleges <- read.csv("../../Data/colleges.csv")

In [11]:
head(colleges)

UnitID Name                                      InState1213 OutOfState1213
1 151351 Indiana University-Bloomington            23116       44566         
2 171100 Michigan State University                 24028       43986         
3 147767 Northwestern University                   60840       60840         
4 204796 Ohio State University-Main Campus         24919       40327         
5 214777 Pennsylvania State University-Main Campus 31854       44156         
6 243780 Purdue University-Main Campus             23468       42270         
  GradRate
1 75      
2 79      
3 93      
4 82      
5 86      
6 69

In summarize, we can calculate any number of measures. Here, we'll use R's built in `mean` and `median` functions to calculate ... well, you get the idea. 

In [13]:
colleges %>%
  summarise(
    count = n(),
    instatemean = mean(InState1213),
    outstatemean = mean(OutOfState1213),
    instatemedian = median(InState1213),
    outstatemedian = median(OutOfState1213),
  )

count instatemean outstatemean instatemedian outstatemedian
1 14    27652.86    42821.5      24473.5       42194

## Assignment

We're going to put it all together now. We're going to calculate the mean and median salaries of job titles at the University of Nebraska-Lincoln. 

Answer this question:

**What are the top median salaries by job title at UNL? And how does that compare to the average salary for that position?**

To do this, you'll need to [download this data](https://www.dropbox.com/s/yqweytz0sb6cd5f/nusalaries1718.csv?dl=0). You'll also need an additional bit of code: filtering. You can add filter condition using this example: `filter(classes, prefix == "JOUR")` where classes is the dataset, prefix is the column you want to filter and `==` is the equal sign.  

#### Rubric

1. Did you read the data into a dataframe? 
2. Did you use group by syntax correctly? 
3. Did you use summarize syntax correctly?
4. Did you use filter syntax correctly?
5. Did you use Markdown comments to explain your steps? 